internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -264,11 +299,56 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 			req.Input = in
 		}
 
-		// TODO(negz): Generate a content-addressable tag for this request.
-		// Perhaps using https://github.com/cerbos/protoc-gen-go-hashpb ?
-		rsp, err := c.pipeline.RunFunction(ctx, fn.FunctionRef.Name, req)
-		if err != nil {
-			return CompositionResult{}, errors.Wrapf(err, errFmtRunPipelineStep, fn.Step)
+		// used to store the requirements returned at the previous iteration
+		var requirements *v1beta1.Requirements
+		// used to store the response of the function at the previous iteration
+		var rsp *v1beta1.RunFunctionResponse
+
+		for i := int64(0); i <= MaxRequirementsIterations; i++ {
+			// if we reached the maximum number of iterations we need to return an error
+			if i == MaxRequirementsIterations {
+				return CompositionResult{}, errors.Errorf(errFmtFunctionMaxIterations, fn.Step, MaxRequirementsIterations)
+			}
+
+			// TODO(negz): Generate a content-addressable tag for this request.
+			// Perhaps using https://github.com/cerbos/protoc-gen-go-hashpb ?
+			rsp, err = c.pipeline.RunFunction(ctx, fn.FunctionRef.Name, req)
+			if err != nil {
+				return CompositionResult{}, errors.Wrapf(err, errFmtRunPipelineStep, fn.Step)
+			}
+
+			if c.composite.ExtraResourcesGetter == nil {
+				// If we don't have an extra resources getter, we don't need to
+				// iterate to satisfy the requirements.
+				break
+			}
```

Comment:
I think this would only be true if the feature flag was disabled, right? If so, consider this a note that it could be removed if we decide to drop the distinct feature flag for extra resources.

Reply:
yes, only in that case indeed.